### PR TITLE
Create massive action to add and remove a domain to a computer

### PR DESCRIPTION
**Pull request by Lainow**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #15749

### <samp>🤖 Generated by Copilot at 628a2f7</samp>

This pull request adds and improves massive actions for domains in the `Domain` and `Computer` classes. It also updates the API and the tests to support the new actions. It allows users to manage domain relations more easily and consistently.

### CODE REVIEW

**Comment by trasher** [comment]
Could you see if it's possible to add tests? There is already a MA used in tests for transfer; it should not be too difficult to add cases for the new two actions.

**Comment by stonebuzz** [comment]
Here are the requested TU (I don't have the rights to update the PR)

```patch
diff --git a/tests/functional/MassiveAction.php b/tests/functional/MassiveAction.php
index 4eae12bcd7..6e27af9cac 100644
--- a/tests/functional/MassiveAction.php
+++ b/tests/functional/MassiveAction.php
@@ -38,6 +38,7 @@ namespace tests\units;
 use CommonDBTM;
 use Contract;
 use DbTestCase;
+use Domain_Item;
 use Notepad;
 use Problem;
 use Session;
@@ -703,4 +704,124 @@ class MassiveAction extends DbTestCase
             User::class
         );
     }
+
+
+
+    public function testProcessMassiveActionsForOneItemtype_AddDomain()
+    {
+        $this->login('glpi', 'glpi');
+
+        $domain_item = new \Domain_Item();
+
+        //create computer
+        $computer = new \Computer();
+        $this->integer($computer->add([
+            'name' => 'test',
+            'entities_id' => 1,
+        ]))->isGreaterThan(0);
+
+        //create manual domain
+        $manual_domain = new \Domain();
+        $this->integer($manual_domain->add([
+            'name' => 'manual_domain',
+            'entities_id' => 1,
+        ]))->isGreaterThan(0);
+
+
+        $this
+        ->boolean(boolval(Session::haveRight(Domain_Item::$rightname, UPDATE)))
+        ->isIdenticalTo(true);
+
+        // Execute action to link Computer and Manual Domain
+        $this->processMassiveActionsForOneItemtype(
+            "add_item",
+            $computer,
+            [$computer->fields['id']],
+            ['domains_id' => $manual_domain->fields['id'], 'domainrelations_id' => \DomainRelation::BELONGS],
+            1,
+            0,
+            \Domain::class
+        );
+
+        //check if exist
+        $rows = $domain_item->find([
+            'domains_id'            => $manual_domain->fields['id'],
+            'items_id'              => $computer->fields['id'],
+            'itemtype'              => $computer->getType(),
+            'domainrelations_id'    => \DomainRelation::BELONGS,
+            'is_deleted'            => false,
+        ]);
+        $this->integer(count($rows))->isEqualTo(1);
+
+
+        //create new domain (for is_dynamic test)
+        $dynamic_domain = new \Domain();
+        $this->integer($dynamic_domain->add([
+            'name' => 'dynamic_domain',
+        ]))->isGreaterThan(0);
+
+        //add relation (with is_dynamic = 1)
+        $this->integer($domain_item->add([
+            'domains_id'            => $dynamic_domain->fields['id'],
+            'is_dynamic'            => 1,
+            'items_id'              => $computer->fields['id'],
+            'itemtype'              => $computer->getType(),
+            'domainrelations_id'    => \DomainRelation::BELONGS,
+            'is_deleted'            => false,
+        ]))->isGreaterThan(0);
+
+
+        // Execute action to remove link between Computer and Manual Domain
+        $this->processMassiveActionsForOneItemtype(
+            "remove_domain",
+            $computer,
+            [$computer->fields['id']],
+            ['domains_id' => $manual_domain->fields['id']],
+            1,
+            0,
+            \Domain::class
+        );
+
+        //manual Domain link should not exist
+        $rows = $domain_item->find([
+            'domains_id'            => $manual_domain->fields['id'],
+            'items_id'              => $computer->fields['id'],
+            'itemtype'              => $computer->getType(),
+            'domainrelations_id'    => \DomainRelation::BELONGS,
+        ]);
+        $this->integer(count($rows))->isEqualTo(0);
+
+        //dynamic domain still exist
+        $rows = $domain_item->find([
+            'domains_id'            => $dynamic_domain->fields['id'],
+            'is_dynamic'            => 1,
+            'items_id'              => $computer->fields['id'],
+            'itemtype'              => $computer->getType(),
+            'domainrelations_id'    => \DomainRelation::BELONGS,
+            'is_deleted'            => false,
+        ]);
+        $this->integer(count($rows))->isEqualTo(1);
+
+        // Execute action to remove link between Computer and Dynamic Domain
+        $this->processMassiveActionsForOneItemtype(
+            "remove_domain",
+            $computer,
+            [$computer->fields['id']],
+            ['domains_id' => $dynamic_domain->fields['id']],
+            1,
+            0,
+            \Domain::class
+        );
+
+        //dynamic domain still exist but in trashbin
+        $rows = $domain_item->find([
+            'domains_id'            => $dynamic_domain->fields['id'],
+            'is_dynamic'            => 1,
+            'items_id'              => $computer->fields['id'],
+            'itemtype'              => $computer->getType(),
+            'domainrelations_id'    => \DomainRelation::BELONGS,
+            'is_deleted'            => true,
+        ]);
+        $this->integer(count($rows))->isEqualTo(1);
+    }
 }

```

**Comment by trasher** [comment]
> Here are the requested TU (I don't have the rights to update the PR)

I've added it, thanks :)

**Review by trasher — APPROVED** [review]

**Review by stonebuzz — APPROVED** [review]

File: tests/functional/MassiveAction.php [hunk 1]
```diff
@@ -703,4 +704,124 @@ public function testProcessMassiveActionsForOneItemtype_deleteEmail(
...
+            "remove_domain",
+            $computer,
+            [$computer->fields['id']],
+            ['domains_id' => $dynamic_domain->fields['id']],
+            1,
+            0,
+            \Domain::class
+        );
+
+        //dynamic domain still exist but in trashbin
```
Comment (cedric-anne) [inline tests/functional/MassiveAction.php:816]:
```suggestion
        //dynamic domain still exist but in trashbin and locked
```

File: src/Domain.php [hunk 2]
```diff
@@ -457,25 +470,55 @@ public static function processMassiveActionsForOneItemtype(MassiveAction $ma, Co
...
+                foreach ($ids as $id) {
+                    $domain_item = new Domain_Item();
+                    foreach (
+                        $domain_item->find([
+                            'domains_id' => $input['domains_id'],
+                            'items_id'   => $id,
+                            'itemtype'   => $item->getType()
+                        ]) as $data
+                    ) {
+                        if ($domain_item->delete($data, !$data['is_dynamic'])) {
```
Comment (cedric-anne) [inline src/Domain.php:510]:
```suggestion
                        $purge = !$data['is_dynamic']; // dynamic relations should be preserved for inventory lock feature (dynamic + deleted = locked)
                        if ($domain_item->delete($data, $purge)) {
```

**Review by cedric-anne — APPROVED** [review]

File: src/Domain.php (hunk 2 above)
Comment (trasher) [reply src/Domain.php]: That seems correct, but I've asked @stonebuzz because I'm not 100% sure

**Pull request merged**